File: app/src/main/java/dev/suhockii/lifetest/ui/details/DetailsPresenter.java
package dev.suhockii.lifetest.ui.details;

import android.view.View;

import com.arellomobile.mvp.InjectViewState;

import javax.inject.Inject;
import javax.inject.Named;

import dev.suhockii.lifetest.model.Product;
import dev.suhockii.lifetest.model.ProductDetails;
import dev.suhockii.lifetest.repo.AppRepository;
import dev.suhockii.lifetest.util.mvp.NonLeakPresenter;
import dev.suhockii.lifetest.util.rx_transformers.RxSchedulers;
import io.reactivex.Single;
import timber.log.Timber;

import static dev.suhockii.lifetest.di.module.RepositoryModule.LOCAL_REPOSITORY;
import static dev.suhockii.lifetest.di.module.RepositoryModule.REMOTE_REPOSITORY;

@InjectViewState
public class DetailsPresenter extends NonLeakPresenter<DetailsView> {

    private final AppRepository localRepository;
    private final AppRepository remoteRepository;
    private final RxSchedulers rxSchedulers;

    @Inject
    public DetailsPresenter(@Named(LOCAL_REPOSITORY) AppRepository localRepository,
                            @Named(REMOTE_REPOSITORY) AppRepository remoteRepository,
                            RxSchedulers rxSchedulers) {
        this.localRepository = localRepository;
        this.remoteRepository = remoteRepository;
        this.rxSchedulers = rxSchedulers;
    }

    @Override
    protected void onFirstViewAttach() {
        super.onFirstViewAttach();
        getViewState().onFirstPresenterAttach();
    }

    void loadDetailsFor(Product product) {
        getViewState().showProgressBar(View.VISIBLE);

        localRepository.getDetailsFor(product)
                .onErrorResumeNext(localEmpty -> updateDetailsFor(product))
                .doOnSubscribe(compositeDisposable::add)
                .compose(rxSchedulers.getIoToMainTransformerSingle())
                .doOnSuccess(productDetails -> getViewState().clearShowSnackbarCommand())
                .doOnError(throwable -> getViewState().showSnackbar(throwable.getMessage(), () -> loadDetailsFor(product)))
                .doOnEvent((productDetails, throwable) -> getViewState().showProgressBar(View.INVISIBLE))
                .subscribe(productDetails -> getViewState().showDetails(productDetails), Timber::e);
    }

    private Single<ProductDetails> updateDetailsFor(Product product) {
        return remoteRepository.getDetailsFor(product)
                .doOnSuccess(localRepository::saveProductDetails);
    }
}

File: app/src/main/java/dev/suhockii/lifetest/ui/details/DetailsPresenter.kt
package dev.suhockii.lifetest.ui.details

import android.view.View

import moxy.InjectViewState

import javax.inject.Inject
import javax.inject.Named

import dev.suhockii.lifetest.model.Product
import dev.suhockii.lifetest.model.ProductDetails
import dev.suhockii.lifetest.repo.AppRepository
import dev.suhockii.lifetest.util.mvp.NonLeakPresenter
import dev.suhockii.lifetest.util.rx_transformers.RxSchedulers
import io.reactivex.Single
import timber.log.Timber

import dev.suhockii.lifetest.di.module.RepositoryModule.LOCAL_REPOSITORY
import dev.suhockii.lifetest.di.module.RepositoryModule.REMOTE_REPOSITORY

@InjectViewState
class DetailsPresenter @Inject constructor(
    @param:Named(LOCAL_REPOSITORY) private val localRepository: AppRepository,
    @param:Named(REMOTE_REPOSITORY) private val remoteRepository: AppRepository,
    private val rxSchedulers: RxSchedulers
) : NonLeakPresenter<DetailsView>() {

    override fun onFirstViewAttach() {
        super.onFirstViewAttach()
        viewState.onFirstPresenterAttach()
    }

    internal fun loadDetailsFor(product: Product) {
        viewState.showProgressBar(View.VISIBLE)

        localRepository.getDetailsFor(product)
            .onErrorResumeNext { updateDetailsFor(product) }
            .doOnSubscribe{ compositeDisposable.add(it) }
            .compose(rxSchedulers.getIoToMainTransformerSingle())
            .doOnSuccess { viewState.clearShowSnackbarCommand() }
            .doOnError { throwable -> viewState.showSnackbar(throwable.message, Runnable { loadDetailsFor(product) }) }
            .doOnEvent { _, throwable ->
                Timber.e(throwable)
                viewState.showProgressBar(View.INVISIBLE) }
            .subscribe { productDetails -> viewState.showDetails(productDetails) }
    }

    private fun updateDetailsFor(product: Product): Single<ProductDetails> {
        return remoteRepository.getDetailsFor(product)
            .doOnSuccess { localRepository.saveProductDetails(it) }
    }
}
